File: Stage32/Python/Problem14002.py
import sys

input = sys.stdin.readline

def main():
    N = int(input())
    arr = list(map(int, input().split()))

    dp = [1] * N
    prev = [-1] * N

    max_len = 0
    max_idx = 0

    for i in range(N):
        for j in range(i):
            if arr[j] < arr[i] and dp[j] + 1 > dp[i]:
                dp[i] = dp[j] + 1
                prev[i] = j
        if dp[i] > max_len:
            max_len = dp[i]
            max_idx = i

    result = []
    cur = max_idx
    while cur != -1:
        result.append(arr[cur])
        cur = prev[cur]

    result.reverse()

    print(max_len)
    print(*result)

if __name__ == "__main__":
    main()

File: Stage32/C/Problem14002.c
#include <stdio.h>

int main(void) {
    int N;
    scanf("%d", &N);

    int arr[1001];
    int dp[1001];
    int prev[1001];

    for (int i = 0; i < N; i++) {
        scanf("%d", &arr[i]);
        dp[i] = 1;
        prev[i] = -1;
    }

    int maxLen = 0;
    int maxIdx = 0;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < i; j++) {
            if (arr[j] < arr[i] && dp[j] + 1 > dp[i]) {
                dp[i] = dp[j] + 1;
                prev[i] = j;
            }
        }
        if (dp[i] > maxLen) {
            maxLen = dp[i];
            maxIdx = i;
        }
    }

    int result[1001];
    int idx = 0;
    int cur = maxIdx;

    while (cur != -1) {
        result[idx++] = arr[cur];
        cur = prev[cur];
    }

    printf("%d\n", maxLen);
    for (int i = idx - 1; i >= 0; i--) {
        printf("%d ", result[i]);
    }

    return 0;
}